### PR TITLE
Unify move method in PinotFS

### PR DESCRIPTION
**Pull request by jackjlli**
This PR unifies move method in different implementations of PinotFS.

In AzureFS, a file cannot move to a non-empty directory. Since the only usage of this API in Pinot is to move a file to a file, we add a note in PinotFS interface to suggest always specifying the names of both src and dst to make less confusion.

### CODE REVIEW

File: pinot-azure-filesystem/src/main/java/org/apache/pinot/filesystem/AzurePinotFS.java [hunk 1]
```diff
@@ -100,6 +100,18 @@ public boolean delete(URI segmentUri, boolean forceDelete)
...
       throws IOException {
+    if (exists(dstUri)) {
+      if (overwrite) {
+        delete(dstUri, true);
+      } else {
+        // dst file exists, returning
+        return false;
+      }
+    } else {
+      URI parentUri = Paths.get(dstUri).toUri();
```
Comment (mcvsubbu) [inline pinot-azure-filesystem/src/main/java/org/apache/pinot/filesystem/AzurePinotFS.java]: Am i reading this right that we are essentially creating dstUri as a directory but asking the user to pass the input param a a file?

Comment (jackjlli) [reply pinot-azure-filesystem/src/main/java/org/apache/pinot/filesystem/AzurePinotFS.java]: Correct the method according to the latest API doc we made in PinotFS interface.

File: pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/SegmentDeletionManager.java [hunk 2]
```diff
@@ -188,12 +188,13 @@ protected void removeSegmentFromStore(String tableNameWithType, String segmentId
       try {
         if (pinotFS.exists(fileToMoveURI)) {
           // Overwrites the file if it already exists in the target directory.
-          pinotFS.move(fileToMoveURI, deletedSegmentDestURI, true);
-          // Updates last modified.
-          // Touch is needed here so that removeAgedDeletedSegments() works correctly.
-          pinotFS.touch(deletedSegmentDestURI);
-          LOGGER.info("Moved segment {} from {} to {}", segmentId, fileToMoveURI.toString(),
-              deletedSegmentDestURI.toString());
+          if (pinotFS.move(fileToMoveURI, deletedSegmentDestURI, true)) {
```
Comment (sunithabeeram) [inline pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/SegmentDeletionManager.java]: Should an message be logged if move returns false? If we can rely on move to always log, then you can remove the additional log in PinotLLCRealtimeSegmentManager.java

File: pinot-azure-filesystem/src/main/java/org/apache/pinot/filesystem/AzurePinotFS.java [hunk 3]
```diff
@@ -100,9 +100,28 @@ public boolean delete(URI segmentUri, boolean forceDelete)
...
+        } else {
+          // dst file exists, returning
+          return false;
+        }
+      }
+    }
+    // ensures the parent path of dst exists.
+    URI parentUri = Paths.get(dstUri).toUri();
+    _adlStoreClient.createDirectory(parentUri.getPath());
     //rename the file
```
Comment (sunithabeeram) [inline pinot-azure-filesystem/src/main/java/org/apache/pinot/filesystem/AzurePinotFS.java]: This comment is misleading as its assuming a 'file' whereas all of the above code seems more flexible.

File: pinot-hadoop-filesystem/src/main/java/org/apache/pinot/filesystem/HadoopPinotFS.java [hunk 4]
```diff
@@ -92,9 +92,25 @@ public boolean delete(URI segmentUri, boolean forceDelete)
   @Override
   public boolean move(URI srcUri, URI dstUri, boolean overwrite)
       throws IOException {
-    if (exists(dstUri) && !overwrite) {
+    if (!exists(srcUri)) {
```
Comment (sunithabeeram) [inline pinot-hadoop-filesystem/src/main/java/org/apache/pinot/filesystem/HadoopPinotFS.java]: This part of the code exists in all implementations. Can we add this to PinotFS, call super.move() and then handle FS specific logic in the impl classes?

File: pinot-filesystem/src/main/java/org/apache/pinot/filesystem/PinotFS.java [hunk 5]
```diff
@@ -59,12 +59,15 @@ public abstract boolean delete(URI segmentUri, boolean forceDelete)
...
-   * If src is a file and dst is a directory, src file will get moved under dst directory.
+   * If src is a file and dst is an empty directory, src file will get moved under dst directory.
    * If both src and dst are directories, src directory will get moved under dst directory.
    * If src is a directory and dst is a file, operation will fail.
    * For example, if a file /a/b/c is moved to a file /x/y/z, in the case of overwrite, the directory /a/b still exists,
    * but will not contain the file 'c'. Instead, /x/y/z will contain the contents of 'c'.
    * If a file /a is moved to a directory /x/y, all the original files under /x/y will be kept.
+   *
+   * As of now, the usage of this API in Pinot is just to move a file to a file. So please specify the full names of
+   * both src and dst to make less confusion.
```
Comment (sunithabeeram) [inline pinot-filesystem/src/main/java/org/apache/pinot/filesystem/PinotFS.java]: Remove "to make less confusion"

**Comment by mcvsubbu** [comment]
I think we should insist that the interfaces work only if full paths are specified in both src and dest. We can change the usage to be that way, and the implementations will get simpler across file systems.

So, for example it is illegal to move(src, dst)  unless the last part of src and dst match. Same with copy.

@sunithabeeram  what if we approach it this way? We can still support all combinations of moving from file to dir, dir to dir, etc. just restricting the usage.

File: pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/SegmentDeletionManager.java (hunk 2 above)
Comment (jackjlli) [reply pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/SegmentDeletionManager.java]: I think it's essential to log when returning false because the false result may be returned from the deep inner logic of different FS.

File: pinot-azure-filesystem/src/main/java/org/apache/pinot/filesystem/AzurePinotFS.java (hunk 3 above)
Comment (jackjlli) [reply pinot-azure-filesystem/src/main/java/org/apache/pinot/filesystem/AzurePinotFS.java]: Adjusted the comment.

File: pinot-filesystem/src/main/java/org/apache/pinot/filesystem/PinotFS.java (hunk 5 above)
Comment (jackjlli) [reply pinot-filesystem/src/main/java/org/apache/pinot/filesystem/PinotFS.java]: Done.

**Comment by sunithabeeram** [comment]
@mcvsubbu since the common logic is moved to the abstract class now, its a bit more easier to enforce the contracts across FS impls. We have noted exceptions in case of AzureFS - didn't want to twist the top level interface because one FS behaved differently.

File: pinot-filesystem/src/main/java/org/apache/pinot/filesystem/PinotFS.java [hunk 6]
```diff
@@ -22,14 +22,19 @@
...
+import java.nio.file.Paths;
 import org.apache.commons.configuration.Configuration;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 
 /**
  * The PinotFS is intended to be a thin wrapper on top of different filesystems. This interface is intended for internal
  * Pinot use only. This class will be implemented for each pluggable storage type.
  */
```
Comment (sunithabeeram) [inline pinot-filesystem/src/main/java/org/apache/pinot/filesystem/PinotFS.java]: Lets add notes on some of the core discussions we have around this so its clear to other why the interface is the way it is:
"PinotFS is a restricted FS API that exposes functionality that is required for Pinot to use different FS implementations. The restrictions are in place due to 2 driving factors:
- Prevent unexpected performance hit when a broader API is implemented - especially, we would like to reduce calls to remote filesystems that might be needed for a broader API, but not necessarily required by Pinot (see the documentation for move() method below).
- Provide an interface that is simple to be implemented across different FS types. The contract that developers have to adhere to will be simpler.

Please read the method level docs carefully to note the exceptions while using the APIs."

File: pinot-filesystem/src/main/java/org/apache/pinot/filesystem/PinotFS.java [hunk 7]
```diff
@@ -58,31 +63,53 @@ public abstract boolean delete(URI segmentUri, boolean forceDelete)
...
    * that haven't been created, this method will create all the necessary parent directories.
-   * If both src and dst are files, dst will be overwritten.
-   * If src is a file and dst is a directory, src file will get moved under dst directory.
-   * If both src and dst are directories, src directory will get moved under dst directory.
-   * If src is a directory and dst is a file, operation will fail.
+   * Note: In Pinot we recommend the full paths of both src and dst be specified.
    * For example, if a file /a/b/c is moved to a file /x/y/z, in the case of overwrite, the directory /a/b still exists,
    * but will not contain the file 'c'. Instead, /x/y/z will contain the contents of 'c'.
-   * If a file /a is moved to a directory /x/y, all the original files under /x/y will be kept.
+   * If a directory /a/b/ is renamed to another directory /x/y/, the directory /x/y/ will contains the content of /a/b/.
```
Comment (sunithabeeram) [inline pinot-filesystem/src/main/java/org/apache/pinot/filesystem/PinotFS.java]: Lets not mix terminology here: Give examples in terms of move API so its clear to readers. i.e., if srcUri = /a/b; an the contents of /a/b are: /a/b/c, /a/b/d; and dstUri=/x/y, the result will be /a (directory b removed), and dst will be /x/y/c, /x/y/d.

And then provide the exception case: if you would like to move b under /x/y, invoke move with srcUri=/a/b and dstUri=/x/y/b.

File: pinot-filesystem/src/main/java/org/apache/pinot/filesystem/PinotFS.java [hunk 8]
```diff
@@ -58,31 +63,53 @@ public abstract boolean delete(URI segmentUri, boolean forceDelete)
...
    * that haven't been created, this method will create all the necessary parent directories.
-   * If both src and dst are files, dst will be overwritten.
-   * If src is a file and dst is a directory, src file will get copied under dst directory.
-   * If both src and dst are directories, src directory will get copied under dst directory.
-   * If src is a directory and dst is a file, operation will fail.
-   * For example, if a file /x/y/z is copied to /a/b/c, /x/y/z will be retained and /x/y/z will also be present as /a/b/c;
-   * if a file /a is copied to a directory /x/y, all the original files under /x/y will be kept.
+   * Note: In Pinot we recommend the full paths of both src and dst be specified.
+   * For example, if a file /a/b/c is copied to a file /x/y/z, in the case of overwrite, the directory /a/b still exists,
+   * but will not contain the file 'c'. Instead, /x/y/z will contain the contents of 'c'.
```
Comment (sunithabeeram) [inline pinot-filesystem/src/main/java/org/apache/pinot/filesystem/PinotFS.java]: That seems wrong. Source will still have c. Its just that if /x/y/z had a c, c will be overwritten.

File: pinot-filesystem/src/main/java/org/apache/pinot/filesystem/PinotFS.java [hunk 9]
```diff
@@ -58,31 +63,53 @@ public abstract boolean delete(URI segmentUri, boolean forceDelete)
...
+   * If a directory /a/b/ is copied to another directory /x/y/, the directory /x/y/ will contains the content of /a/b/.
```
Comment (sunithabeeram) [inline pinot-filesystem/src/main/java/org/apache/pinot/filesystem/PinotFS.java]: Might be good to rephrase this as in the move example with src and dst arguments used for the calls.

File: pinot-controller/src/main/java/org/apache/pinot/controller/ControllerConf.java [hunk 10]
```diff
@@ -170,8 +170,8 @@ public static URI getUriFromPath(String path) {
 
   public static URI constructSegmentLocation(String baseDataDir, String tableName, String segmentName) {
     try {
-      return new URI(StringUtil.join(File.separator, baseDataDir, tableName, URLEncoder.encode(segmentName, "UTF-8")));
-    } catch (UnsupportedEncodingException | URISyntaxException e) {
+      return getUriFromPath(StringUtil.join(File.separator, baseDataDir, tableName, URLEncoder.encode(segmentName, "UTF-8")));
```
[the remark this answers is not in the document]
Comment (jackjlli) [reply pinot-controller/src/main/java/org/apache/pinot/controller/ControllerConf.java]: Yes, this change is needed as `creating parent directory` logic is put to base method level in PinotFS class.

File: pinot-hadoop-filesystem/src/main/java/org/apache/pinot/filesystem/HadoopPinotFS.java (hunk 4 above)
Comment (jackjlli) [reply pinot-hadoop-filesystem/src/main/java/org/apache/pinot/filesystem/HadoopPinotFS.java]: Common logic has been moved to base class.

File: pinot-filesystem/src/main/java/org/apache/pinot/filesystem/PinotFS.java (hunk 6 above)
Comment (jackjlli) [reply pinot-filesystem/src/main/java/org/apache/pinot/filesystem/PinotFS.java]: Thanks for the detail! I'll add your comments to the API doc.

File: pinot-filesystem/src/main/java/org/apache/pinot/filesystem/PinotFS.java (hunk 8 above)
Comment (jackjlli) [reply pinot-filesystem/src/main/java/org/apache/pinot/filesystem/PinotFS.java]: Good catch! 👍 Addressed.

File: pinot-filesystem/src/main/java/org/apache/pinot/filesystem/PinotFS.java (hunk 9 above)
Comment (jackjlli) [reply pinot-filesystem/src/main/java/org/apache/pinot/filesystem/PinotFS.java]: Tests added for this two scenario.

File: pinot-filesystem/src/main/java/org/apache/pinot/filesystem/PinotFS.java (hunk 7 above)
Comment (jackjlli) [reply pinot-filesystem/src/main/java/org/apache/pinot/filesystem/PinotFS.java]: Comments updated.

File: pinot-azure-filesystem/src/main/java/org/apache/pinot/filesystem/AzurePinotFS.java [hunk 11]
```diff
@@ -98,12 +98,13 @@ public boolean delete(URI segmentUri, boolean forceDelete)
   }
 
   @Override
-  public boolean move(URI srcUri, URI dstUri, boolean overwrite)
+  /**
```
[the remark this answers is not in the document]
Comment (jackjlli) [reply pinot-azure-filesystem/src/main/java/org/apache/pinot/filesystem/AzurePinotFS.java]: Removed.

File: pinot-filesystem/src/main/java/org/apache/pinot/filesystem/PinotFS.java [hunk 12]
```diff
@@ -58,31 +69,54 @@ public abstract boolean delete(URI segmentUri, boolean forceDelete)
...
    * Copies the file or directory from the src to dst. The original file is retained. If the dst has parent directories
    * that haven't been created, this method will create all the necessary parent directories.
-   * If both src and dst are files, dst will be overwritten.
-   * If src is a file and dst is a directory, src file will get copied under dst directory.
-   * If both src and dst are directories, src directory will get copied under dst directory.
-   * If src is a directory and dst is a file, operation will fail.
-   * For example, if a file /x/y/z is copied to /a/b/c, /x/y/z will be retained and /x/y/z will also be present as /a/b/c;
-   * if a file /a is copied to a directory /x/y, all the original files under /x/y will be kept.
+   * Note: In Pinot we recommend the full paths of both src and dst be specified.
+   * For example, if a file /a/b/c is copied to a file /x/y/z, in the case of overwrite, the directory /a/b still exists,
```
Comment (sunithabeeram) [inline pinot-filesystem/src/main/java/org/apache/pinot/filesystem/PinotFS.java]: Didn't catch this earlier - I don't see an overwrite for copy! We can move addressing/reviewing copy related changes out of this PR if you prefer.

Comment (jackjlli) [reply pinot-filesystem/src/main/java/org/apache/pinot/filesystem/PinotFS.java]: There's no code changes for copy in this PR. So no harm to leave the doc change here. :)

Comment (sunithabeeram) [reply pinot-filesystem/src/main/java/org/apache/pinot/filesystem/PinotFS.java]: Lets not have incorrect documentation though, please :)

Comment (jackjlli) [reply pinot-filesystem/src/main/java/org/apache/pinot/filesystem/PinotFS.java]: The doc has been updated without mentioning `overwrite`.
